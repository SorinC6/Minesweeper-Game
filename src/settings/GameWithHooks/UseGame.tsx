import { useEffect, useState } from 'react';

import { Field, emptyFieldGenerator, CellState, fieldGenerator, Coords } from '@/helpers/Filed';
import { openCell } from '@/helpers/openCell';

import { LevelNames, GameSettings } from '../GameSettings';
import { setFlag } from '@/helpers/setFlag';

interface ReturnType {
  level: LevelNames;
  isGameOver: boolean;
  isWin: boolean;
  settings: [number, number];
  playerField: Field;
  onClick: (coords: Coords) => void;
  onContextMenu: (coords: Coords) => void;
  onChangeLevel: (level: LevelNames) => void;
  isGameStarted: boolean;
  onReset: () => void;
  gameField: Field;
  time: number;
  flagCounter: number;
}

export const useGame = (): ReturnType => {
  const [level, setLevel] = useState<LevelNames>('beginner');
  const [isGameOver, setIsGameOver] = useState(false);
  const [isWin, setIsWin] = useState(false);
  const [time, setTime] = useState(0);
  const [flagCounter, setFlagCounter] = useState(0);
  const [isGameStarted, setIsGameStarted] = useState(false);

  const [size, bombs] = GameSettings[level];
  const [playerField, setPlayerField] = useState<Field>(emptyFieldGenerator(size, CellState.hidden));
  const [gameField, setGameField] = useState<Field>(fieldGenerator(size, bombs / (size * size)));

  console.log('gameField', gameField);

  useEffect(() => {
    let interval: NodeJS.Timeout;

    if (isGameStarted) {
      interval = setTimeout(() => {
        setTime(time + 1);
      }, 1000);
      if (isGameOver) {
        clearInterval(interval);
      }
    }

    return () => {
      clearInterval(interval);
    };
  }, [isGameOver, isGameStarted, time]);

  const setGameOver = (isSolved = false) => {
    setIsGameOver(true);
    setIsWin(isSolved);
  };

  const onClick = (coords: Coords) => {
    !isGameStarted && setIsGameStarted(true);
    try {
      const [newPlayerField, isSolved] = openCell(coords, playerField, gameField);
      if (isSolved) {
        setGameOver(isSolved);
      }
      setPlayerField([...newPlayerField]);
    } catch (error) {
      setPlayerField([...gameField]);
      setGameOver();
    }
  };

  const onChangeLevel = (level: LevelNames) => {
    setLevel(level);
    const newSettings = GameSettings[level];
    resetHandle(newSettings);
  };

  const resetHandle = ([size, bombs]: [number, number]) => {
    const newGameField = fieldGenerator(size, bombs / (size * size));
    const newPlayerField = emptyFieldGenerator(size, CellState.hidden);
    setGameField([...newGameField]);
    setPlayerField([...newPlayerField]);
    setIsGameOver(false);
    setIsWin(false);
    setIsGameStarted(false);
    setTime(0);
  };

  const onReset = () => resetHandle([size, bombs]);

  const onContextMenu = (coords: Coords) => {
    !isGameStarted && setIsGameStarted(true);
    const [newPlayerField, isSolved, newFlagCounter] = setFlag(coords, playerField, gameField, flagCounter, bombs);
    setFlagCounter(newFlagCounter);
    if (isSolved) {
      setGameOver(isSolved);
    }
    setPlayerField([...newPlayerField]);
  };

  return {
    time,
    flagCounter,
    level,
    isGameOver,
    isWin,
    settings: [size, bombs],
    playerField,
    onClick,
    onChangeLevel,
    onReset,
    gameField,
    onContextMenu,
    isGameStarted,
  };
};
