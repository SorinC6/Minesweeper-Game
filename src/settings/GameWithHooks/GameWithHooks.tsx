import React, { FC } from 'react';

import { Grid } from '@/components/Grid';

import { Top } from '@/components/Top';
import { Scoreboard } from '@/components/Scoreboard';
import { GameArea, Wrapper, GameOver } from '@/components/Game';

import { GameLevels, LevelNames } from '../GameSettings';

import { useGame } from '../../helpers/UseGame';

export const GameWithHooks: FC = () => {
  const { level, isGameOver, isWin, settings, playerField, onClick, onChangeLevel, onReset } = useGame();

  const [, bombs] = settings;

  return (
    <Wrapper>
      <Top feature="Flag" firstAction="ctrl" secondAction="click">
        Minesweeper
      </Top>
      <GameArea>
        <Scoreboard
          time="000"
          levels={GameLevels as unknown as string[]}
          mines={String(bombs)}
          onReset={onReset}
          defaultLevel={level}
          onChangeLevel={({ target: { value: level } }) => {
            onChangeLevel(level as LevelNames);
          }}
        />
        {isGameOver && <GameOver onClick={() => null} isWin={isWin} />}
        <Grid onClick={onClick} onContextMenu={() => null}>
          {playerField}
        </Grid>
      </GameArea>
    </Wrapper>
  );
};
