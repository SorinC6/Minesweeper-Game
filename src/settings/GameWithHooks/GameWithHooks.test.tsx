import React from 'react';
import userEvent from '@testing-library/user-event';
import { render, screen, fireEvent } from '@testing-library/react';

import { GameWithHooks } from './GameWithHooks';

const mockOnClick = jest.fn();
const mockOnChangeLevel = jest.fn();
const mockOnReset = jest.fn();
const mockOnContextMenu = jest.fn();
const mockSetSearchParams = jest.fn();

describe('GAME with hooks tests', () => {
  test('Render game field by default', () => {
    const { asFragment } = render(<GameWithHooks />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('Change level works fine', () => {
    render(<GameWithHooks />);
    userEvent.selectOptions(screen.getByRole('combobox'), 'intermediate');
    // expect(mockOnChangeLevel).toHaveBeenCalled();
    // expect(mockSetSearchParams).toHaveBeenCalledWith({ level: 'intermediate' });
  });
});
