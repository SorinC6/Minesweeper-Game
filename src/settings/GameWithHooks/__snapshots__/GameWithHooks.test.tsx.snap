// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameWithHooks test cases Render game field by default 1`] = `
<DocumentFragment>
  <div
    class="css-1ct4owg"
  >
    <header
      class="css-dn2347"
    >
      <h1
        class="css-18kvaxm"
      >
        Minesweeper
      </h1>
      <legend
        class="css-17qz6a2"
      >
        <strong>
          Flag: 
        </strong>
        <code
          class="css-12ap3z0"
        >
          <span
            class="css-qm6gfa"
          >
            ctrl
          </span>
           + 
          <span
            class="css-l2ia8y"
          >
            click
          </span>
        </code>
      </legend>
    </header>
    <div
      class="css-1bh93n8"
    >
      <div
        class="css-1n7y7g1"
      >
        <div
          class="css-d2q1pa"
        >
          000
        </div>
        <div>
          <select
            class="css-1lpmfg0"
            data-testid="select"
            role="combobox"
          >
            <option
              class="css-zcmdxr"
              value="beginner"
            >
              beginner
            </option>
            <option
              class="css-zcmdxr"
              value="intermediate"
            >
              intermediate
            </option>
            <option
              class="css-zcmdxr"
              value="expert"
            >
              expert
            </option>
          </select>
          <button
            class="css-1bva717"
          >
            🙂
          </button>
        </div>
        <div
          class="css-d2q1pa"
        >
          10
        </div>
      </div>
      <div
        class="css-1p4hvro"
      >
        🙁
      </div>
      <div
        class="css-da0epy"
        size="2"
      >
        <div
          class="css-707t19"
          data-testid="0,0"
          role="cell"
        >
          10
        </div>
        <div
          class="css-707t19"
          data-testid="0,1"
          role="cell"
        >
          10
        </div>
        <div
          class="css-707t19"
          data-testid="1,0"
          role="cell"
        >
          10
        </div>
        <div
          class="css-707t19"
          data-testid="1,1"
          role="cell"
        >
          10
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
